mountpoint-s3/src/upload.rs

```diff
@@ -31,88 +32,234 @@ where
     #[error("put request failed")]
     PutRequestFailed(#[from] E),
 
     #[error("out of order write; expected offset {expected_offset:?} but got {write_offset:?}")]
     OutOfOrderWrite { write_offset: u64, expected_offset: u64 },
+
+    #[error("put request had already completed")]
+    PutRequestAlreadyCompleted,
+
+    #[error("put request had previously failed")]
+    PutRequestPreviouslyFailed,
 }
 
 /// Manages the upload of an object to S3.
 ///
-/// Wraps a PutObject request and enforces sequential writes.
-pub struct UploadRequest<Client: ObjectClient> {
-    bucket: String,
+/// Handles the lifecycle of a PutObject request,
+/// invalidates it on errors, and enforces sequential writes.
+#[derive(Debug)]
+pub struct UploadRequest<Client: ObjectClient, Handle> {
     key: String,
     next_request_offset: u64,
-    request: Client::PutObjectRequest,
+    state: UploadRequestState<Client, Handle>,
+}
+
+enum UploadRequestState<Client: ObjectClient, Handle> {
+    InProgress {
+        request: Client::PutObjectRequest,
+        handle: Handle,
+    },
+    Completed,
+    Failed,
 }
 
-impl<Client> UploadRequest<Client>
+impl<Client, Handle> UploadRequest<Client, Handle>
 where
     Client: ObjectClient + Send + Sync + 'static,
 {
     async fn new(
         inner: Arc<UploaderInner<Client>>,
         bucket: &str,
         key: &str,
+        handle: Handle,
     ) -> ObjectClientResult<Self, PutObjectError, Client::ClientError> {
         let request = inner
             .client
             .put_object(bucket, key, &PutObjectParams::default())
             .await?;
 
         Ok(Self {
-            bucket: bucket.to_owned(),
             key: key.to_owned(),
             next_request_offset: 0,
-            request,
+            state: UploadRequestState::InProgress { request, handle },
         })
     }
 
     pub fn size(&self) -> u64 {
         self.next_request_offset
     }
 
-    pub async fn write(&mut self, offset: i64, data: &[u8]) -> Result<(), UploadWriteError<PutRequestError<Client>>> {
+    pub fn is_in_progress(&self) -> bool {
+        matches!(self.state, UploadRequestState::InProgress { .. })
+    }
+
+    pub async fn write(&mut self, offset: i64, data: &[u8]) -> Result<usize, UploadError<PutRequestError<Client>>> {
         let next_offset = self.next_request_offset;
         if offset != next_offset as i64 {
-            return Err(UploadWriteError::OutOfOrderWrite {
+            return Err(UploadError::OutOfOrderWrite {
                 write_offset: offset as u64,
                 expected_offset: next_offset,
             });
         }
 
-        self.request.write(data).await?;
-        self.next_request_offset += data.len() as u64;
-        Ok(())
+        let request = match &mut self.state {
+            UploadRequestState::InProgress { request, .. } => request,
+            UploadRequestState::Completed => {
+                error!(key = self.key, "object already uploaded");
```

Comment:
See #310 -- for internal stuff (everything below the `fs.rs`), we should try to keep `error` logging out of cases where it's an "expected failure".